### PR TITLE
Rename Blazor Templates(#10348)

### PR DESCRIPTION
**Pull request by isaacrlevin**
Summary of the changes (Less than 80 chars)
 - Rename RazorComponents project to blazorserver
 - Rename Standalone Blazor project to blazorwasm with option `--host` that will add ASP.NET Core hosted backend

Addresses #10348, #11700

### CODE REVIEW

**Comment by isaac2004** [comment]
@danroth27 here is the ServerSide rename work, blocked on wasm template

**Comment by ryanbrandenburg** [comment]
@aspnet-hello triage https://dev.azure.com/dnceng/public/_build/results?buildId=238021

**Comment by aspnet-hello** [comment]
This comment was made automatically. If there is a problem contact aspnetcore-build@microsoft.com.

I've triaged the above build. I've created/commented on the following issue(s)
https://github.com/aspnet/AspNetCore-Internal/issues/2718

**Comment by isaac2004** [comment]
@danroth27 	here are the client-side updates. Will need a review on my approach with templates, but they build and work.

**Comment by isaac2004** [comment]
@phenning one thing that I may need some help with is moving the contents of a folder up one level using the template engine

**Comment by ryanbrandenburg** [comment]
@isaac2004 looks like this will be good to go once we get a review from @danroth27 and a rebase to solve these conflicts.

File: src/Components/Blazor/Templates/src/content/BlazorHosted-CSharp/Client/Pages/FetchData.razor
Comment (ryanbrandenburg) [inline src/Components/Blazor/Templates/src/content/BlazorHosted-CSharp/Client/Pages/FetchData.razor]:
```suggestion
        forecasts = await Http.GetJsonAsync<WeatherForecast[]>("WeatherForecast");
```

File: src/Components/Blazor/Templates/src/content/BlazorHosted-CSharp/Server/Controllers/SampleDataController.cs
Comment (ryanbrandenburg) [inline src/Components/Blazor/Templates/src/content/BlazorHosted-CSharp/Server/Controllers/SampleDataController.cs]:
```suggestion
    [Route("[controller]")]
```

File: src/Components/Blazor/Templates/src/content/BlazorHosted-CSharp/Server/Controllers/SampleDataController.cs
Comment (ryanbrandenburg) [inline src/Components/Blazor/Templates/src/content/BlazorHosted-CSharp/Server/Controllers/SampleDataController.cs]:
```suggestion
    public class WeatherForecastController : Controller
```

File: src/Components/Blazor/Templates/src/content/BlazorHosted-CSharp/Server/Controllers/SampleDataController.cs
Comment (ryanbrandenburg) [inline src/Components/Blazor/Templates/src/content/BlazorHosted-CSharp/Server/Controllers/SampleDataController.cs]:
```suggestion
        public WeatherForecastController(ILogger<WeatherForecastController> logger)
        {
            this.logger = logger;
        }
```

File: src/Components/Blazor/Templates/src/content/BlazorHosted-CSharp/Server/Controllers/SampleDataController.cs
Comment (ryanbrandenburg) [inline src/Components/Blazor/Templates/src/content/BlazorHosted-CSharp/Server/Controllers/SampleDataController.cs]:
```suggestion
        [HttpGet]
```

File: src/Components/Blazor/Templates/src/content/BlazorHosted-CSharp/Server/Controllers/SampleDataController.cs
Comment (ryanbrandenburg) [inline src/Components/Blazor/Templates/src/content/BlazorHosted-CSharp/Server/Controllers/SampleDataController.cs]:
```suggestion
        public IEnumerable<WeatherForecast> Get()
```

File: src/Components/Blazor/Templates/src/content/BlazorLibrary-CSharp/.template.config.src/template.json
Comment (ryanbrandenburg) [inline src/Components/Blazor/Templates/src/content/BlazorLibrary-CSharp/.template.config.src/template.json]: What happened to these files?

File: src/Components/Blazor/Templates/src/content/BlazorStandalone-CSharp/.template.config.src/dotnetcli.host.json
Comment (ryanbrandenburg) [inline src/Components/Blazor/Templates/src/content/BlazorStandalone-CSharp/.template.config.src/dotnetcli.host.json]: These files too?

File: src/ProjectTemplates/Web.ProjectTemplates/content/BlazorServerWeb-CSharp/.template.config/template.json
Comment (ryanbrandenburg) [inline src/ProjectTemplates/Web.ProjectTemplates/content/BlazorServerWeb-CSharp/.template.config/template.json]:
```suggestion
  "name": "Blazor Server App",
```

File: src/ProjectTemplates/Web.ProjectTemplates/content/RazorComponentsWeb-CSharp/Pages/Counter.razor
Comment (ryanbrandenburg) [inline src/ProjectTemplates/Web.ProjectTemplates/content/RazorComponentsWeb-CSharp/Pages/Counter.razor]: More missing files? Did I miss some reason that we should have removed these?

**Review by ryanbrandenburg — CHANGES_REQUESTED** [review]

File: src/Components/Blazor/Templates/src/content/BlazorHosted-CSharp/Hosted-CSharp.sln
Comment (ryanbrandenburg) [inline src/Components/Blazor/Templates/src/content/BlazorHosted-CSharp/Hosted-CSharp.sln]: If this csproj only exists conditionally then this project definition should also be conditional.

Comment (isaac2004) [reply src/Components/Blazor/Templates/src/content/BlazorHosted-CSharp/Hosted-CSharp.sln]: Question, do we need a solution with 1 project? The goal here is to have either a generated solution with 1 or 3 depending on "host" option

**Review by danroth27 — CHANGES_REQUESTED** [review]
Looking good, @isaac2004! Thanks for contributing all this work!

We need to figure out if we can get rid of the "Client" folder and solution file when `hosted` isn't specified. @phenning ?

**Comment by isaac2004** [comment]
fixed some js reference issues per #11700

File: src/ProjectTemplates/Web.ProjectTemplates/content/RazorPagesWeb-CSharp/Pages/Shared/_Layout.cshtml
Comment (ryanbrandenburg) [inline src/ProjectTemplates/Web.ProjectTemplates/content/RazorPagesWeb-CSharp/Pages/Shared/_Layout.cshtml]: Let's not mix concerns here. Generally we try to keep PR's to addressing one issue because it makes them easier to review.

File: src/Components/Blazor/Templates/src/content/BlazorHosted-CSharp/.template.config.src/template.json
Comment (danroth27) [inline src/Components/Blazor/Templates/src/content/BlazorHosted-CSharp/.template.config.src/template.json]: These conditions are necessary to ensure the template works property in VS.

**Review by danroth27 — CHANGES_REQUESTED** [review]

File: src/ProjectTemplates/Web.ProjectTemplates/content/RazorPagesWeb-CSharp/Pages/Shared/_Layout.cshtml
Comment (isaac2004) [reply src/ProjectTemplates/Web.ProjectTemplates/content/RazorPagesWeb-CSharp/Pages/Shared/_Layout.cshtml]: no worries, want me to revert?

Comment (ryanbrandenburg) [reply src/ProjectTemplates/Web.ProjectTemplates/content/RazorPagesWeb-CSharp/Pages/Shared/_Layout.cshtml]: Yes please.

Comment (isaac2004) [reply src/ProjectTemplates/Web.ProjectTemplates/content/RazorPagesWeb-CSharp/Pages/Shared/_Layout.cshtml]: done

File: src/Components/Blazor/Templates/src/content/BlazorHosted-CSharp/Hosted-CSharp.sln
Comment (isaac2004) [reply src/Components/Blazor/Templates/src/content/BlazorHosted-CSharp/Hosted-CSharp.sln]: So in the template, I leave out the sln, so it won't be created if hosted = false

File: src/Components/Blazor/Templates/src/content/BlazorHosted-CSharp/.template.config.src/template.json
Comment (isaac2004) [reply src/Components/Blazor/Templates/src/content/BlazorHosted-CSharp/.template.config.src/template.json]: Yea, I was unsure how to set the output. I have a fix for it

File: src/Components/Blazor/Templates/src/content/BlazorHosted-CSharp/.template.config.src/template.json
Comment (isaac2004) [reply src/Components/Blazor/Templates/src/content/BlazorHosted-CSharp/.template.config.src/template.json]: I will add back

**Comment by isaac2004** [comment]
@ryanbrandenburg @danroth27 worked with @phenning to complete the authoring work. Templates are working through CLI and VS (including 16.3.1 option work). Can you and possibly @SteveSandersonMS review the template themselves?

@ryanbrandenburg also let me know if there are merge conflicts I need to do on my side, I moved the above files so there are conflicts that I manually added

**Comment by ryanbrandenburg** [comment]
I can't resolve these merge conflicts from the UI. When possible please resolve any merge conflicts you're aware of as it seems that the UI is pretty picky about what it will allow and not having push access to your fork makes it impossible to for me to resolve them on this PR.

**Comment by isaac2004** [comment]
@ryanbrandenburg resolved the merge on my side. Should be good if it passes tests and  review

**Review by ryanbrandenburg — APPROVED** [review]

File: src/Components/Blazor/Templates/src/content/BlazorHosted-CSharp/.template.config.src/template.json
Comment (danroth27) [inline src/Components/Blazor/Templates/src/content/BlazorHosted-CSharp/.template.config.src/template.json]: I'm not sure how classifications are used, but should "Hosted" be replaced with "WebAssembly"?

File: src/Components/Blazor/Templates/src/content/BlazorHosted-CSharp/.template.config.src/template.json
Comment (danroth27) [inline src/Components/Blazor/Templates/src/content/BlazorHosted-CSharp/.template.config.src/template.json]: Should we rename "BlazorHosted" to "BlazorWasm" everywhere?

File: src/Components/Blazor/Templates/src/content/BlazorHosted-CSharp/Client/BlazorHosted-CSharp.Client.csproj
Comment (danroth27) [inline src/Components/Blazor/Templates/src/content/BlazorHosted-CSharp/Client/BlazorHosted-CSharp.Client.csproj]: Is there any reason to not always include this so that you have the option to run the client project directly?

File: src/Components/Blazor/Templates/src/content/BlazorHosted-CSharp/Client/Pages/Counter.razor
Comment (danroth27) [inline src/Components/Blazor/Templates/src/content/BlazorHosted-CSharp/Client/Pages/Counter.razor]: In Preview 7 the second `@` isn't needed.

```suggestion
<button class="btn btn-primary" @onclick="IncrementCount">Click me</button>
```

File: src/Components/Blazor/Templates/src/content/BlazorHosted-CSharp/Client/Shared/NavMenu.razor
Comment (danroth27) [inline src/Components/Blazor/Templates/src/content/BlazorHosted-CSharp/Client/Shared/NavMenu.razor]:
```suggestion
<div class="@NavMenuCssClass" @onclick="ToggleNavMenu">
```

**Review by danroth27 — CHANGES_REQUESTED** [review]
Looks really good! Thanks for doing all this work! Just a few comments.

File: src/Components/Blazor/Templates/src/content/BlazorHosted-CSharp/Client/BlazorHosted-CSharp.Client.csproj
Comment (isaac2004) [reply src/Components/Blazor/Templates/src/content/BlazorHosted-CSharp/Client/BlazorHosted-CSharp.Client.csproj]: Goal was to have parity between what was already there. If the above is valuable, no issue there

File: src/Components/Blazor/Templates/src/content/BlazorHosted-CSharp/.template.config.src/template.json
Comment (isaac2004) [reply src/Components/Blazor/Templates/src/content/BlazorHosted-CSharp/.template.config.src/template.json]: sure

File: src/Components/Blazor/Templates/src/content/BlazorHosted-CSharp/.template.config.src/template.json
Comment (isaac2004) [reply src/Components/Blazor/Templates/src/content/BlazorHosted-CSharp/.template.config.src/template.json]: ok

File: src/Components/Blazor/Templates/src/content/BlazorHosted-CSharp/Client/Shared/NavMenu.razor
Comment (isaac2004) [reply src/Components/Blazor/Templates/src/content/BlazorHosted-CSharp/Client/Shared/NavMenu.razor]: ok

File: src/Components/Blazor/Templates/src/content/BlazorHosted-CSharp/Client/Pages/Counter.razor
Comment (isaac2004) [reply src/Components/Blazor/Templates/src/content/BlazorHosted-CSharp/Client/Pages/Counter.razor]: ok

**Comment by isaac2004** [comment]
@danroth27 changes complete

**Comment by isaac2004** [comment]
On a side note, the following docs will need to be updated after merging

https://docs.microsoft.com/en-us/aspnet/core/host-and-deploy/blazor/client-side?view=aspnetcore-3.0
https://docs.microsoft.com/en-us/aspnet/core/blazor/hosting-models?view=aspnetcore-3.0
https://docs.microsoft.com/en-us/aspnet/core/blazor/get-started?view=aspnetcore-3.0&tabs=visual-studio
https://docs.microsoft.com/en-us/aspnet/core/security/blazor/?view=aspnetcore-3.0&tabs=visual-studio

File: src/ProjectTemplates/Web.ProjectTemplates/content/BlazorServerWeb-CSharp/Shared/NavMenu.razor
Comment (danroth27) [inline src/ProjectTemplates/Web.ProjectTemplates/content/BlazorServerWeb-CSharp/Shared/NavMenu.razor]:
```suggestion
    <button class="navbar-toggler" @onclick="ToggleNavMenu">
```

File: src/Components/Blazor/Templates/src/content/BlazorWasm-CSharp/.template.config.src/vs-2017.3.host.json
Comment (danroth27) [inline src/Components/Blazor/Templates/src/content/BlazorWasm-CSharp/.template.config.src/vs-2017.3.host.json]:
```suggestion
    "text": "Blazor WebAssembly App",
```

File: src/ProjectTemplates/Web.ProjectTemplates/content/BlazorServerWeb-CSharp/.template.config/vs-2017.3.host.json
Comment (danroth27) [inline src/ProjectTemplates/Web.ProjectTemplates/content/BlazorServerWeb-CSharp/.template.config/vs-2017.3.host.json]: Update text above from "Blazor (server-side)" to "Blazor Server App".

**Review by danroth27 — CHANGES_REQUESTED** [review]
A couple more small things. Otherwise looks good!

**Comment by isaac2004** [comment]
@phenning can you confirm the vs template shows the Hosted option after Dan's requested changes?

File: src/Components/Blazor/Templates/src/content/BlazorWasm-CSharp/Client/BlazorWasm-CSharp.Client.csproj
Comment (danroth27) [inline src/Components/Blazor/Templates/src/content/BlazorWasm-CSharp/Client/BlazorWasm-CSharp.Client.csproj]: Fix indentation?

**Review by danroth27 — APPROVED** [review]
Looks good to me.

**Comment by isaac2004** [comment]
@danroth27 can you confirm the VS dialog is what you want
image

image

**Comment by danroth27** [comment]
@isaac2004 If you're willing, I'd like to make two more tweaks:

1. Update the Blazor WebAssembly App description to say: A project template for creating a Blazor app that runs on WebAssembly **and is optionally hosted by an ASP.NET Core app**. This template can be used for web apps with rich dynamic user interfaces (UIs).

2. Update the text for the "Hosted" checkbox to say "ASP.NET Core hosted".

**Comment by isaac2004** [comment]
@danroth27 both requests are complete

**Comment by danroth27** [comment]
:shipit:

**Review by ryanbrandenburg — APPROVED** [review]
It's a big PR but I gave it another gander and it looks good. We've got to wait for the tests again after the rebase, but I'll merge it once I see the green.

**Comment by ryanbrandenburg** [comment]
The failures here seem to be substantive. 

- _Host.cshtml file needs to contain a "BOM" character.
- `Startup.cs(15,48): error CS0234: The type or namespace name 'Areas' does not exist in the namespace 'AspNet.blazorserverindividualuld._7cbeb0' (are you missing an assembly reference?)`

**Comment by isaac2004** [comment]
@ryanbrandenburg what is going on with this startup file

src/ProjectTemplates/Web.ProjectTemplates/content/BlazorServerWeb-CSharp/Startup.cs

It seems like it doesn't like the condition at the bottom. No matter what I do, it does not seem to build the template right. I think that is the issue that is happening

**Comment by isaac2004** [comment]
@ryanbrandenburg failed flaky tests but CI passed

**Comment by ryanbrandenburg** [comment]
The build is technically green, but you still have [this failure](https://dev.azure.com/dnceng/public/_build/results?buildId=260589&view=ms.vss-test-web.build-test-results-tab&runId=6974916&resultId=100022&paneView=debug) that I mentioned above. It didn't fail the build because it's marked flaky, but in this case it's an actual break and it needs to be fixed. The problem seems to be that the "Areas/**" path is not being created, so it fails when they're referenced in Startup.cs.

**Comment by isaac2004** [comment]
@danroth27 party time!

**Pull request merged**
